### PR TITLE
Use partial but correct vtable layout

### PR DESCRIPTION
**Pull request by oli-obk**
r? @RalfJung who suggested to also do this change for nightly, not just beta

### CODE REVIEW

**Comment by RalfJung** [comment]
You only added that more detailed layout specifically for validation, right? So this should not regress anything else?

Please add a comment in `layout.rs` why we do not have a more detailed layout (in particular, why the "obvious thing" you did is wrong).

**Comment by oli-obk** [comment]
There's a `FIXME` right below said line.

Yea, only validation and debug symbols were affected

**Comment by RalfJung** [comment]
That FIXME doesn't tell me much -- in particular, it doesn't explain that even the *size* of the vtable, naively computed, is wrong (the thing that lead to this having to be backported, right?)

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -1627,6 +1627,13 @@ impl<'a, 'tcx, C> TyLayoutMethods<'tcx, C> for Ty<'tcx>
                             tcx.mk_array(tcx.types.usize, 3),
                         )
                         /* FIXME use actual fn pointers
+                        Warning: naively computing the number of entries in the
+                        vtable by counting the methods on the trait + methods on
+                        all parent traits does not work, because some methods can
+                        be not object safe and thus excluded from the vtable.
+                        Increase this counter if you tried to implement this but
+                        failed to do it without duplicating a lot of code from
+                        other places in the compiler: 2
```
Comment (RalfJung) [inline src/librustc/ty/layout.rs]: :rofl:

**Review by eddyb — APPROVED** [review]

**Pull request merged**